tracing/ops_tracing/buffer.py

```diff
@@ -0,0 +1,298 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+# file except in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+"""Buffer for tracing data in the ops-tracing extension."""
+
+from __future__ import annotations
+
+import contextlib
+import functools
+import logging
+import sqlite3
+from pathlib import Path
+from typing import Callable
+
+from typing_extensions import ParamSpec, TypeVar
+
+from .const import Config
+
+# Approximate safety limit for the database file size
+BUFFER_SIZE = 40 * 1024 * 1024
```

Comment:
Similar to the comment in the constants file elsewhere: do we expect charmers to use these? If not, could this module be `_buffer.py` instead, or these constants private?

Either way, if they're not meant to change, which the all-caps seems to indicate, then `typing.Final` would be good I think.

Reply:
TIL about `typing.Final`. I reckon if they're internal it's not worth marking them final, as it would just annoy us if we have to override them in tests (I think?).

Reply:
I'd rather not sprinkle Final around.
There are places in OpenTelemetry where I wish they had not.